### ISSUE TITLE
Feature request: S3 Backend

### ISSUE THREAD

**Issue by fdefilippo** [drakkan/sftpgo#60]
It would be very convenient to have the ability to use the credentials: accesskey & secretkey of an s3 storage and then connect to the storage itself as backend.

**Comment by drakkan** [comment]
Hi, thanks for your suggestion. I plan to analyze cloud storage backends (S3 and others) after 0.9.5 release that I'm finalizing.
Currently I don't have access to a S3 backend, I'll initially use the one year free plan, after the first year to be able to maintain the feature someone should give me access to a test account

**Comment by jovandeginste** [comment]
@drakkan you may nee interested in minio, a self hosted s3 server, for testing purposes

**Comment by fdefilippo** [comment]
@drakkan you have many option:

1) use https://play.minio.io:9000 
2) download minio, (one file) and run as server (https://docs.min.io/docs/minio-quickstart-guide.html)
3) I can give you access to my s3 endpoint.

**Comment by drakkan** [comment]
@jovandeginste @fdefilippo I'll try minio, thanks!

**Comment by drakkan** [comment]
Hi,

I have a working proof of concept implementation that can upload/download files hosted on minio via SFTP.

To properly integrate this I need to abstract the filesystem calls, it will require a while, anyway I'll try to implement this feature after 0.9.5 release.

I plan to assign a bucket to each sftp user, the assigned bucket is the equivalent of the root directory for the local users. Is this ok for your use case? 

I noted that aws provides a managed SFTP transfer that use S3 as backend, 

can you please better explain your use case and why someone should prefer SFTPGo (that needs to be installed somewhere) to this aws service? I'm quite new to aws and I could miss something obvious, is the aws SFTP service very expensive? Are there other reasons? Thanks

**Comment by fdefilippo** [comment]
hello, today I use rclone (http://rclone.org) using it as an sftp server and using the auth-proxy interface, very convenient because it allows me to map users on the proper s3 nodes. I found your server and seeing the features (single session, quota and so on) I would prefer to use it, also I would like to replace some proftpd servers (with the sftp module) with yours. 

As for the s3 backend, how do you intend to map users?
- using the administrative keys to then access the buckets that users have access to? (it might make sense in the case of ssh keys)
- using credentials (username / password) as s3 credentials?

**Comment by drakkan** [comment]
> hello, today I use rclone (http://rclone.org) using it as an sftp server and using the auth-proxy interface, very convenient because it allows me to map users on the proper s3 nodes.

I'll try this rclone mode, thanks for sharing the info

 I found your server and seeing the features (single session, quota and so on) I would prefer to use it, also I would like to replace some proftpd servers (with the sftp module) with yours.
> 
> As for the s3 backend, how do you intend to map users?
> 
> * using the administrative keys to then access the buckets that users have access to? (it might make sense in the case of ssh keys)
> * using credentials (username / password) as s3 credentials?

I plan to let you add users as now, so you can define SFTP users with password and/or public keys. This credential are for the SFTP server (SFTPGo in this case). I plan to add a new user field that let you specify the user type (local or cloud) and for cloud users you can define the cloud details such as the bucket and the cloud credentials. 

So you login with SFTP users as usual and you can upload/download and list files stored in s3/minio based on the cloud configuration associated to the user. 

Basically the remote or local filesystem is served via SFTP based on the user type

> 
> ps: I don't use aws. I'm using minio on internal network for custom projects.

**Comment by drakkan** [comment]
Hi,

we have basic S3 support now, I plan to add more S3 features in the coming weeks,

please let me know if it works for you and if you suggestions, thanks!

**Comment by fdefilippo** [comment]
Hi, great work as expected! 

 

Thank you!

 

f.

 

Da: drakkan <notifications@github.com> 
Inviato: domenica 19 gennaio 2020 07:47
A: drakkan/sftpgo <sftpgo@noreply.github.com>
Cc: fdefilippo <francescodefilippo+github@gmail.com>; Mention <mention@noreply.github.com>
Oggetto: Re: [drakkan/sftpgo] Feature request: S3 Backend (#60)

 

Hi,

we have basic S3 support now, I plan to add more S3 features in the coming weeks,

please let me know if it works for you and if you suggestions, thanks!
